pkg/controllers/v1alpha1/dataload/implement.go

```diff
@@ -1,482 +1,384 @@
 }
 
-// Return a new reconciler implement
-func NewReconcilerImplement(client client.Client, log logr.Logger, recorder record.EventRecorder) *ReconcilerImplement {
-	r := &ReconcilerImplement{
+// NewDataLoadReconcilerImplement returns a DataLoadReconcilerImplement
+func NewDataLoadReconcilerImplement(client client.Client, log logr.Logger, recorder record.EventRecorder) *DataLoadReconcilerImplement {
+	r := &DataLoadReconcilerImplement{
 		Client:   client,
 		Log:      log,
 		Recorder: recorder,
 	}
 	return r
 }
 
-// Set up the reconciler
-func (r *ReconcilerImplement) Setup() {
-	if r.DataLoaderImage != "" {
-		return
-	}
+// ReconcileDataLoadDeletion reconciles the deletion of the DataLoad
+func (r *DataLoadReconcilerImplement) ReconcileDataLoadDeletion(ctx reconcileRequestContext) (ctrl.Result, error) {
+	log := ctx.Log.WithName("ReconcileDataLoadDeletion")
 
-	val, existed := os.LookupEnv(common.ENV_DATALOADER_IMG)
-	if !existed {
-		r.DataLoaderImage = common.DATALOAD_DEFAULT_IMAGE
-	} else {
-		r.DataLoaderImage = val
+	// 1. Delete release if exists
+	releaseName := utils.GetDataLoadReleaseName(ctx.DataLoad.Name)
+	err := helm.DeleteReleaseIfExists(releaseName, ctx.DataLoad.Namespace)
+	if err != nil {
+		log.Error(err, "can't delete release", "releaseName", releaseName)
+		return utils.RequeueIfError(err)
 	}
-	r.Log.Info("Setting up DataLoader Image", "Image", r.DataLoaderImage)
-}
 
-// Reconcile Dataload
-func (r *ReconcilerImplement) ReconcileDataload(ctx cdataload.ReconcileRequestContext) (ctrl.Result, error) {
-	/*
-		DataLoadPhase: Complete
-	*/
-	if ctx.DataLoad.Status.Phase == common.DataloadPhaseComplete {
-		return r.reconcileCompleteDataload(ctx)
+	// 2. Release lock on target dataset if necessary
+	err = r.releaseLockOnTargetDataset(ctx, log)
+	if err != nil {
+		log.Error(err, "can't release lock on target dataset", "targetDataset", ctx.DataLoad.Spec.Dataset)
+		return utils.RequeueIfError(err)
 	}
 
-	/*
-		DataLoadPhase: Failed
-	*/
-	if ctx.DataLoad.Status.Phase == common.DataloadPhaseFailed {
-		return r.reconcileFailedDataload(ctx)
+	// 3. remove finalizer
+	if utils.HasDeletionTimestamp(ctx.DataLoad.ObjectMeta) {
+		ctx.DataLoad.ObjectMeta.Finalizers = utils.RemoveString(ctx.DataLoad.ObjectMeta.Finalizers, cdataload.DATALOAD_FINALIZER)
+		if err := r.Update(ctx, &ctx.DataLoad); err != nil {
+			log.Error(err, "failed to remove finalizer")
+			return utils.RequeueIfError(err)
+		}
+		log.Info("Finalizer is removed")
 	}
+	return utils.NoRequeue()
+}
 
-	/*
-		DataLoadPhase: None -> Pending
-	*/
-	if ctx.DataLoad.Status.Phase == common.DataloadPhaseNone {
-		return r.reconcileNoneDataload(ctx)
-	}
+// ReconcileDataLoad reconciles the DataLoad according to its phase status
+func (r *DataLoadReconcilerImplement) ReconcileDataLoad(ctx reconcileRequestContext) (ctrl.Result, error) {
+	log := ctx.Log.WithName("ReconcileDataLoad")
+	log.V(1).Info("process the cdataload", "cdataload", ctx.DataLoad)
```

Comment:
Suggest to change from if else to switch.

Reply:
Or add comments, this logic is not easy to understand

Reply:
Thanks. Fixed